### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [11]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
total_players = pd.DataFrame({"Total Players": [len(purchase_data["SN"].unique())]})
total_players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
unique_items = len(purchase_data["Item ID"].unique())
number_of_purchases = purchase_data["Price"].count()
total_revenue = purchase_data["Price"].sum()
average_price = total_revenue / number_of_purchases

summary_df = pd.DataFrame({"Number_of_Unique_Items": [unique_items],
                              "Average_Price": [average_price],
                              "Number_of_Purchases": [number_of_purchases],
                              "Total_Revenue": [total_revenue]})

summary_df["Total_Revenue"] = summary_df["Total_Revenue"].map("${:,.2f}".format)
summary_df["Average_Price"] = summary_df["Average_Price"].map("${:.2f}".format)

summary_df

Number_of_Unique_Items Average_Price  Number_of_Purchases Total_Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
grouping_heroes=purchase_data.groupby('Gender')['SN'].nunique()
total_players = len(purchase_data["SN"].unique())
hero_purchase_data = grouping_heroes.to_frame('Total_Count')
hero_purchase_data['Percentage_of_Players'] = round(hero_purchase_data/total_players*100,2)
hero_purchase_data

Total_Count  Percentage_of_Players
Gender                                                   
Female                          81                  14.06
Male                           484                  84.03
Other / Non-Disclosed           11                   1.91


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
purchase_count = purchase_data.groupby('Gender')['Item ID'].count()
purchase_value = purchase_data.groupby('Gender')['Price'].sum()
purchase_analysis_df=purchase_count.to_frame('Purchase_Count')

purchase_analysis_df["Average_Purchase_Price"] = (purchase_value/purchase_count).map("${:.2f}".format)
purchase_analysis_df["Total_Purchase_Value"] = purchase_value.map("${:.2f}".format)
purchase_analysis_df["Average_Total_Purchase_per_Person"] = (purchase_value/grouping_heroes).map("${:.2f}".format)
purchase_analysis_df

Purchase_Count Average_Purchase_Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total_Purchase_Value Average_Total_Purchase_per_Person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                              $1967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
age_bins = [0,9, 14, 19, 24, 29, 34, 39, max(purchase_data['Age'])]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29","30-34","35-39","40+"]
purchase_data['Age_Bracket'] = pd.cut(purchase_data["Age"], age_bins, labels=group_names)
players_count = purchase_data.groupby('Age_Bracket')['SN'].nunique()
age_demos_df = players_count.to_frame('Total_Count').sort_index()
age_demos_df['Percentage_of_Players']=round(age_demos_df['Total_Count']/total_players * 100,2)
age_demos_df

Total_Count  Percentage_of_Players
Age_Bracket                                    
<10                   17                   2.95
10-14                 22                   3.82
15-19                107                  18.58
20-24                258                  44.79
25-29                 77                  13.37
30-34                 52                   9.03
35-39                 31                   5.38
40+                   12                   2.08

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
purchase_age_analysis_df = purchase_data.groupby('Age_Bracket')['Purchase ID'].count().to_frame('Purchase_Count').sort_index()

purchase_age_analysis_df['Average_Purchase_Price'] = purchase_data.groupby('Age_Bracket')['Price'].mean().map("${:.2f}".format)

total_purchase_value = purchase_data.groupby('Age_Bracket')['Price'].sum()


purchase_age_analysis_df['Avg_Total_Purchase_per_Person'] = (total_purchase_value /players_count).map("${:.2f}".format)
purchase_age_analysis_df['Total_Purchase_Value'] = total_purchase_value .map("${:.2f}".format)

purchase_age_analysis_df

Purchase_Count Average_Purchase_Price  \
Age_Bracket                                          
<10                      23                  $3.35   
10-14                    28                  $2.96   
15-19                   136                  $3.04   
20-24                   365                  $3.05   
25-29                   101                  $2.90   
30-34                    73                  $2.93   
35-39                    41                  $3.60   
40+                      13                  $2.94   

            Avg_Total_Purchase_per_Person Total_Purchase_Value  
Age_Bracket                                                     
<10                                 $4.54               $77.13  
10-14                               $3.76               $82.78  
15-19                               $3.86              $412.89  
20-24                               $4.32             $1114.06  
25-29                               $3.81              $293.00  
30-34                               $4.12              $214.00  
35-39                               $4.76              $147.67  
40+                                 $3.19               $38.24

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
purchase_count=purchase_data.groupby('SN')['Purchase ID'].nunique()
top_spenders_df = purchase_count.to_frame('Purchase_Count')

top_spenders = purchase_data.groupby('SN')['Price'].sum()
top_spenders_df['Average_Purchase_Price'] = (top_spenders/top_spenders_df['Purchase_Count']).map("${:.2f}".format)
top_spenders_df['Total_Purchase_Value'] = top_spenders

value_df = top_spenders_df.sort_values(by='Total_Purchase_Value', ascending=False)
value_df['Total_Purchase_Value'] = value_df['Total_Purchase_Value'].map("${:.2f}".format)
value_df.head()

Purchase_Count Average_Purchase_Price Total_Purchase_Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
most_popular_items = purchase_data.filter(['Item ID', 'Item Name','Price'], axis=1)
item_amount = most_popular_items.groupby(['Item ID','Item Name'])['Price'].mean()
purchase_count = most_popular_items.groupby('Item ID')['Item Name'].value_counts()
purchase_value = item_amount * purchase_count

most_popular_items_df = purchase_count.to_frame('Purchase_Count')
most_popular_items_df['Item_Price'] = item_amount.map("${:.2f}".format)
most_popular_items_df['Total_Purchase_Value'] = purchase_value
most_popular_items_dfA = most_popular_items_df.sort_values(by="Purchase_Count", ascending=False)
most_popular_items_dfA['Total_Purchase_Value'] = most_popular_items_dfA['Total_Purchase_Value'].map("${:.2f}".format)
most_popular_items_dfA.head()


Purchase_Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item_Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total_Purchase_Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
most_profitable_itemsB = most_popular_items_df.sort_values(by="Total_Purchase_Value", ascending=False)
most_profitable_itemsB['Total_Purchase_Value'] = most_profitable_itemsB['Total_Purchase_Value'].map("${:.2f}".format)
most_profitable_itemsB.head()

Purchase_Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item_Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total_Purchase_Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

Homework Analysis - You must include a written description of at least two observable trends based on the data.

1.) Of the 576 players, the majority of the players who played this free game were Male and also spent the most money on the free optional items that were available in the game. The purchase price for males was much lower in average compared to the females, this may be due to the multiple purchases made over time through the life of the gameplay. Possibly, the type of game being a fantasy game involving heroes may imply that this could be an action/fighter game. The audience may not be directed toward the more feminine clients. 


2.) The top age bracket that purchased this free game was 20-24 and makes up 44.79% compared to the other listed age brackets. What is interesting to see is that this group does not have the highest average purchase price/average total per person. This is likely due to the group having more individuals within it to break down the cost. What you can see is this group spent much more in free optional items. This bracket may flourish due to the ability to purchase free items because of their own income in comparison to the younger age bracket. This 20-24 age bracket may be considered "young" and may not have obligations that pull from their free available time in comparison to older age brackets. 
